Azure/azure-functions-python-worker
# Append /home/site/wwwroot to sys.path

### Content In This PR
Create a unittest friendly folder structure: https://github.com/Azure/azure-functions-python-worker/issues/469
1. Append **AzureWebJobsScriptRoot** to sys.path in Core-Tools and VSCode environment
2. Append **function_app_directory** to sys.path when doing specialization in Linux Consumption
3. (No action on Linux Dedicated / Premium since /home/site/wwwroot is already in sys.path)
4. Update unittests
    - Rename brokenimplicit to implicit_import since we now support `import a as b` statement
    - Add a case for testing name collision with implicit import **name_collision**, the worker should pick third-party package.
    - Add a case for testing name collision with \_\_app\_\_ import **name_collision_app_import**, the worker should pick the package from customer's code.
   - Add a case for testing third-party package import **absolute_thirdparty**
   - Add a case for importing non-existing module **module_not_found**
### Description

The current Python function app folder structure suggested in https://docs.microsoft.com/en-us/azure/azure-functions/functions-reference-python#folder-structure is confusing:
1. It introduces a virtual namespace `__app__` which is not a Python standard. This makes Azure Functions Python code less portable.
2. It gives our customer a hard time adding unit tests in Azure Functions.
3. Customer are not able to treat their function triggers as Python module, using absolute import does not work.
```python
from SharedCode import module  # Not working because function project root is not in sys.path
```

We want to enable unit tests and absolute import to our customers, and make our customers code more portable.

### Fix sys.path
We found out the sys.path is inconsistent across Linux Consumption, Linux Dedicated/Premium, and Local Development environment. After we discussed with Brett and Steve, we decide to align these environments by introducing the project root into sys.path.

**Linux Consumption**:
```python
sys.path = [
    "/tmp/functions\\standby\\wwwroot",
    "/home/site/wwwroot/.python_packages/lib/site-packages",
    "/azure-functions-host/workers/python/3.7/LINUX/X64"
]
# This PR appends /home/site/wwwroot during worker specialization
# in _handle__function_environment_reload_request,
# to avoid breaking existing customer apps.
```

**Linux Dedicated/Premium**:
```python
sys.path = [
    "/home/site/wwwroot",
    "/home/site/wwwroot/.python_packages/lib/site-packages",
    "/azure-functions-host/workers/python/3.7/LINUX/X64"
# no action required since /home/site/wwwroot is already there.
]
```

**Local Development using Core Tools and VSCode**:
```python
sys.path = [
    "C:\\Users\\username\\WorkerPath\\python\\3.7\\WINDOWS\\X64",
    "C:\\Users\\username\\ProjectPath\\.venv\\lib\\site-packages"
# append /home/site/wwwroot when the worker starts in worker.py, since customer code already exists and no specialization is required.
]
```

### Message to customer
1. **Customer should not name function trigger with a third-party package name (i.e. Don't name an http trigger **numpy** if you're using numpy package).**
2. We will continue support `__app__` import.
3. To do **absolute import** from SharedCode, you need to add an `__init__.py` file into the SharedCode folder to make it a Python package, and feel free to use `import SharedCode` or `from SharedCode import module` in your function app.
4. To run **unit test**, create a **test** folder along with other triggers. If `__init__.py` file does not exist in your function trigger, add one into the folder to make it a Python package, and you can test and treat this function trigger as a normal function package.

### Attachment
[NextSuggestedFolderStructure.zip](https://github.com/Azure/azure-functions-python-worker/files/4929159/NextSuggestedFolderStructure.zip)


---


### PR information

- [X] There are a small number of commits, each of which has an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [X] If applicable, the PR references the bug/issue that it fixes in the description.
- [X] New Unit tests were added for the changes made and CI is passing.

### Quality of Code and Contribution Guidelines
- [X] I have read the [contribution guidelines](https://github.com/Azure/azure-functions-python-worker/blob/master/.github/CONTRIBUTING.md).
